python/tchannel/tornado/peer.py

```diff
@@ -169,13 +179,22 @@ def request(self, **kwargs):
         :param service_threshold:
             If ``hostport`` was not specified, this specifies the score
             threshold at or below which peers will be ignored.
+        :param black_list:
+            Peers on the black_list won't be chosen.
         """
-        return self.choose(
+        kwargs['peer_group'] = self
+        peer = self.choose(
             hostport=kwargs.pop('hostport', None),
             score_threshold=kwargs.pop('score_threshold', None),
-        ).request(**kwargs)
+            black_list=kwargs.pop('black_list', None),
+        )
 
-    def choose(self, hostport=None, score_threshold=None):
+        if peer:
+            return peer.request(**kwargs)
+        else:
+            return None
```

Comment:
Also, no matching peer is a major issue. We can't simply return None here. Consumers will be chaining a `.send` call after it so they'll get an unexplainable None-related error in their call. You should be throwing a very specific exception if no peer was found.